backend/services/rate_limit_middleware.py

```
@@ -0,0 +1,66 @@
+    connections bypass it entirely, so the ``/ws/...` routes are unaffected.
+    """
+
+    async def dispatch(self, request: Request, call_next):
+        if request.url.path in EXEMPT_PATHS:
+            return await call_next(request)
+
+        identity = _client_identity(request)
+        result = await rate_limiter.check(identity)
+
+        if not result.allowed:
+            logger.warning(f"Rate limit exceeded for {identity} on {request.url.path}")
```

Ruff (0.15.21)

[warning] Logging statement uses f-string

(G004)